### DISCUSSION TITLE
How to use generated "Indentifier" type?

### DISCUSSION THREAD

**Discussion by karloluis** [kristiandupont/kanel#444] [Q&A] [answered]
When Kanel autogenerates the types for tables with relations it produces the following;

```typescript
/** Identifier type for app.Carts */
export type CartsId = string & { __brand: 'CartsId' }

export default interface Carts ...
export interface CartsInitializer ...
export interface CartsMutator ...
```

which then when used always produces a typescript error because the string will never match the object part and viceversa.

How then do I go about using the `CartsId` type to insert a new row?

**Comment by kristiandupont** [comment, marked as answer]
When inserting/updating, you have to "convert" your plain string to a `CartsId` somewhere, simply by casting: `const cartsId = '1' as CartsId;`.

It was recently requested that the branded id's should be optional, which it seems several people prefer. I just find them incredibly useful.

**Reply by karloluis** [reply]
That's curious, I don't seem to understand what the motivation would be behind this. Probably some arcane db stuff.

I tend to prefer using `satisfies` to not "trick" the type system, although this also would fail for the same reasons. On the other hand "casting/converting" the type every time seems a bit cumbersome.

**Reply by karloluis** [reply]
That is to say that `const cartsId: CartsId = '1'` will make typescript go boom since it will never match. Casting with `as` will just bind the `__brand` property to a String "object"?

**Reply by kristiandupont** [reply]
It's a way to simulate [nominal typing](https://www.typescriptlang.org/play#example/nominal-typing) in Typescript.
Consider a function with this signature:
```ts
async function assignInvoiceToCustomer(customerId: number, invoiceId: number) { ... }
```

If you accidentally call it with the parameters swapped, you won't get any compiler error because both are numbers. Now, that's probably not very likely to happen but the thing is that *if* it happens, it's Very Bad™️ -- you will put complete garbage into your database and it can be hard to track down and fix the issue. With two types `CustomerId` and `InvoiceId`, this would cause a compiler error.

Note that the "brand" itself is removed completely from the resulting code, it doesn't do anything. And, any function that expects a number will take id's without complaints, just not the other way around.
